### ISSUE TITLE
//regen uses huge chunk of memory.

### ISSUE THREAD

**Issue by josephverburg** [IntellectualSites/FastAsyncWorldEdit#721]
**[REQUIRED] FastAsyncWorldEdit Configuration Files** 

**/fawe debugpaste**: https://athion.net/ISPaster/paste/view/110396d785ca431c8560c1127f0eabe3

**Required Information**
- FAWE Version Number (`/version FastAsyncWorldEdit`):  1.16-427;478ab16
- Spigot/Paper Version Number (`/version`): git-Paper-246
- Minecraft Version: 1.16.3     

**Describe the bug**
When using //regen on a single chunk the memory usage shoots up with 1.7GB (from ~700MB to 2.4 GB), after second //regen it caps out at 2.9 GB (of the 3 GB). Also the single chunk regen takes about 12 seconds.

The recent changes to //regen indicate it should be very usable, however i had to revert to using snapshots of a cleanly generated world (which was very speedy luckily).

**To Reproduce**
1. Select a chunk using //chunk
2. Execute //regen
3. Select a different chunk using //chunk
4. Execute //regen
5. All the memory usage \o/

**Plugins being used on the server**
See the Fawe debugpaste

**Checklist**:

- [X] I included all information required in the sections above
- [X] I made sure there are no duplicates of this report [(Use Search)](https://github.com/IntellectualSites/FastAsyncWorldEdit/issues?q=is%3Aissue)
- [X] I made sure I am using an up-to-date version of [FastAsyncWorldEdit for 1.16.3](https://ci.athion.net/job/FastAsyncWorldEdit-1.16/)
- [X] I made sure the bug/error is not caused by any other plugin

**Comment by aurorasmiles** [comment]
something like a spark profile would be pretty useful to see whats wrong. I'm not surprised that the regen is taking a bit of memory tho, due to the way it works (it basically generates a new temporary world for the regen)

**Comment by Kadeluxe** [comment]
I got ~10x less regen duration by excluding structures generation.

**Comment by josephverburg** [comment]
I agree, however tripling the server memory however is too extreme. And not getting cleaned up and increasing further after the second time implies a memory leak somewhere.

**Comment by aurorasmiles** [comment]
Thats why a profile or a memory dump would help finding that issue

**Comment by SirYwell** [comment]
> something like a spark profile would be pretty useful to see whats wrong.

I tested regenerating 2048(x256)x2048 selections without experiencing memory issues. 
Structure generation is sadly super bad implemented by minecraft. The whole FAWE-side part of //regen performs very well, although we might need to look into improving speed for small selections.

**Comment by josephverburg** [comment]
@aurorasmiles How would go about making such a memory profile, looking here for the best tool

**Comment by aurorasmiles** [comment]
Again, either use spark or just create a memory dump from the java process

**Comment by josephverburg** [comment]
Okay after some digging in a heapdump i found a huge memory consumption by "ConcurrentHashMap$TreeNode" which is referenced from:

sharedAreaMap in com.sk89q.worldedit.bukkit.adapter.impl.regen.Regen_v1_16_R2$FastWorldGenContextArea#46		48 B (0%)	n/a
and
sharedMap in com.sk89q.worldedit.bukkit.adapter.impl.regen.Regen_v1_16_R2$FastAreaLazy#51		32 B (0%)	n/a

**Comment by SirYwell** [comment]
Could you also provide some information about your //regen, like the region size? What you're seeing is mostly caused by structures but it should be cleaned up afterwards

**Comment by josephverburg** [comment]
This was just //set 0 a single chunk then //regen 4 times (4 x //set 0 ; 4 x //regen)

**Comment by MattBDev** [comment]
If a large chunk of memory is being consumed by a ConcurrentHashMap then that can be fixed.

**Comment by SirYwell** [comment]
I just tested and while the ConcurrentHashMap uses a fair amount of ram, it doesn't look like a memory leak for me. Just because the GC didn't cleaned it up yet doesn't mean the entries are still referenced. Profiling with YourKit only showed me remaining unreachable references that can be cleaned up by the GC at any time the memory is required.

**Comment by MattBDev** [comment]
@SirYwell Did YourKit point out if any of the data in the ConcurrentHashMap is redundant. I believe it can look to check for that.
I already found a way to save memory a little bit but it is extremely tiny that it won't make a noticable difference.

**Comment by MattBDev** [comment]
@josephverburg Could you run a test with WorldEdit and see how the memory usage compares with FAWE?

**Comment by josephverburg** [comment]
@MattBDev Test on normal WorldEdit (WorldEdit 7.2.0 Release Candidate 2 (1.13.2 - 1.16.3)) barely uses any extra memory after 10 regens (maybe 100MB, not more) and also seems to be faster then FAWE.

**Comment by jokbon** [comment]
Plotsquared might be afected too. Multiple plot clear = out of memory error. (noticed in october)

**Comment by dordsor21** [comment]
I can't really see that there's anything outside of GC being conducted less often than when doing world generation (from my testing). Standard world generation seems to create just as much (if not more in many cases) "garbage" as //regen.

Unless someone has some very persuasive proof of there being a genuine memory leak/reason why large amounts of memory cannot be cleared by GC then this can be reopened
